### ISSUE TITLE
Running StreamConsumer with tokio's executor

### ISSUE THREAD

**Issue by pimeys** [fede1024/rust-rdkafka#85]
Currently how StreamConsumer stores the `JoinHandle` in a `Cell` prevents us of using the consumer in tokio's [thread_pool](https://docs.rs/tokio/0.1.6/tokio/executor/thread_pool/index.html) executor, which expects the consumer to implement `Send`. In other hand using the [current_thread](https://docs.rs/tokio/0.1.6/tokio/executor/current_thread/index.html) executor doesn't comply with missing static lifetimes somewhere in the consumer code I haven't been able to locate yet.

Is there plans to support the new executors, especially when the development of `tokio-core` and `futures-cpupool` seems to be stalled in favor of `tokio`?

**Comment by fede1024** [comment]
I plan on working on porting the library to support `tokio` instead of `tokio-core` in the coming weeks. Could you provide an example of `thread_pool` usage that you would like to be supported?

**Comment by fede1024** [comment]
I made the `StreamConsumer` Sync and updated the example to tokio (instead of tokio-core + futures-cpupool). However even with the `StreamConsumer` now being Sync, it's not currently possible to process the stream on a tokio executor on a separate thread, because the `StreamConsumer` (which is borrowed by `MessageStream`) must have`'static` lifetime to outlive the threads in the executor. It only work with the `current_thread` executor (see updated example).

Unfortunately making the `MessageStream` own the `StreamConsumer` wouldn't work either: the `Stream` trait is similar to the `Iterator` trait in that it doesn't allow to bind the lifetime of the object returned by `poll` to the lifetime of the stream itself (in the same way objects returned from iterators aren't bound to the iterator's lifetime, but to the lifetime of the object the iterator is created from).

To summarize: it doesn't seem to be currently possible to use `MessageStream` on the thread pool, because of the lack of scoped threads in tokio and the way the `Stream` trait is defined. It is of course still possible to spawn multiple threads manually and create a separate consumer on each thread (or share one via `Arc`). `current_thread` works, but of course is more limited. Let me know if you manage to make it work in your code.

**Comment by pimeys** [comment]
I'm my libraries getting the static lifetimes working was just by adding `+ 'static` to returned futures. Where are your changes @fede1024, I could take a look?

**Comment by fede1024** [comment]
Changes are in master and the example is [here](https://github.com/fede1024/rust-rdkafka/blob/master/examples/asynchronous_processing.rs).

**Comment by pimeys** [comment]
Tried to fix this and hit the wall, my try is here so you'll see the approach:

https://github.com/fede1024/rust-rdkafka/pull/87

The error I get is 
```
error[E0277]: the trait bound `*mut rdkafka_sys::bindings::linux_64::rd_kafka_message_s: std::marker::Send` is not satisfied in `futures::Then<futures::Lazy<[closure@examples/asynchronous_processing.rs:85:39: 107:14 borrowed_message:rdkafka::message::BorrowedMessage, producer:rdkafka::producer::FutureProducer, output_topic:std::string::String], futures::AndThen<futures::Lazy<[closure@examples/asynchronous_processing.rs:87:40: 90:20 borrowed_message:rdkafka::message::BorrowedMessage], futures::future::PollFn<[closure@examples/asynchronous_processing.rs:87:56: 90:19 borrowed_message:rdkafka::message::BorrowedMessage]>>, futures::Then<rdkafka::producer::DeliveryFuture, std::result::Result<(), tokio_threadpool::BlockingError>, [closure@examples/asynchronous_processing.rs:98:31: 105:26]>,
[closure@examples/asynchronous_processing.rs:92:38: 106:18 producer:rdkafka::producer::FutureProducer, output_topic:std::string::String]>>, futures::FutureResult<(), ()>, [closure@examples/asynchronous_processing.rs:109:46: 109:56]>`
   --> examples/asynchronous_processing.rs:109:13
    |
109 |             tokio::spawn(message_future.then(|_| ok(())));
    |             ^^^^^^^^^^^^ `*mut rdkafka_sys::bindings::linux_64::rd_kafka_message_s` cannot be sent between threads safely
    |
```

Also removing the lifetimes with the consumers while having unsafe and raw pointers around doesn't feel very safe. Could we actually have a safe wrappers here so doing this kind of refactorings would be easier?

**Comment by fede1024** [comment]
You need the lifetime in the `BorrowedMessage` because that's what links the message to the consumer that actually owns the data. Borrowed messages just offer a view inside the buffer of the consumer. I believe that the `BorrowedMessage` is the safe wrapping that you are referring to. It wraps the `*mut rd_kafka_message_s` in such a way that the user won't have to interact with it directly, and that the lifetime of the pointer is correctly bound to the lifetime of the consumer. The rust compiler will automatically make sure that you are using it only when it's correct to do so, making it safe to use.

**Comment by pimeys** [comment]
But if the lifetime is not `'static`, we can't use it in the work-stealing threadpool that's in Tokio. The pool is very useful and super fast in consumers, allowing you to use stuff like `blocking` for heavy operations and so on. But I don't really know how to bend the code here to work with static lifetimes.

**Comment by fede1024** [comment]
Unfortunately we can't use `'static` unless the consumer itself is allocated statically, since `BorrowedMessage`s have a reference to the consumer (because the payload lives in the consumer's memory buffer).

I'm experimenting on a new implementation of the `StreamConsumer` which now also support a `OwnedMessageStream` that returns an owned copy of each message. This should be (slightly) less efficient because it has to copy some data but in practice it shouldn't make much of a difference. Check out the `feature/threadless_futures` branch. I'm actively experimenting with it so the API might change, but you could give it a spin anyway and see if something like this might help your use case.

**Comment by tyranron** [comment]
Just hit the same issue trying to inject `MessageStream` into actix actor's context.

@fede1024 any news on it? plans to merge it? or any workarounds appeared?  
Maybe we can somehow bake (in user code) `StreamConsumer` into the result `Stream` using combinators/`lazy()`?
Or adopt `lazy_static` as a hack?

**Comment by toxicafunk** [comment]
How is this workaround supposed to work? I'm trying to use it on the `run_async_processor` example so:

```
let owned_stream: OwnedConsumerStream = OwnedConsumerStream {
        upstream: OwningHandle::new_with_fn(
            Box::new(consumer)
            ,
              |cons| {
                 cons.start()
             })
    };
```
but I get 

`error[E0599]: no method named `start` found for type `*const rdkafka::consumer::StreamConsumer` in the current scope`

**Comment by tyranron** [comment]
@toxicafunk try

```rust
let owned_stream = OwnedConsumerStream {
    upstream: OwningHandle::new_with_fn(Box::new(consumer), |c| {
        let cf = unsafe { &*c };
        Box::new(cf.start())
    }),
};
```

**Comment by toxicafunk** [comment]
Awesome! That did the trick!

As far as I can understand, we have to coerce the dereferenciation of `c` because `StreamConsumer` does not implement `Deref`, correct? Or am I missing something?

Thanks a ton!

**Comment by tyranron** [comment]
@toxicafunk correct. And the `unsafe` block is safe here as `OwningHandle` takes care about `consumer` being not moved, which allows self-referential struct here, but in price of heap allocations, however.

**Comment by surajpayu** [comment]
if this issue is not fixed, this library becomes unusable. core.run(f) is deprecated and when i try doing
tokio::run(h), it gives me below error
error[E0277]: `std::cell::Cell<std::option::Option<std::thread::JoinHandle<()>>>` cannot be shared between threads safely
    --> src/main.rs:1836:5                                                                                                                                                
     |                                                                                                                                                                    
1836 |     tokio::run(h);                                                                                                                                                 
     |     ^^^^^^^^^^ `std::cell::Cell<std::option::Option<std::thread::JoinHandle<()>>>` cannot be shared between threads safely                                         
     |                                                                                                                                                                    
     = help: within `rdkafka::consumer::StreamConsumer`, the trait `std::marker::Sync` is not implemented for `std::cell::Cell<std::option::Option<std::thread::JoinHandle<()>>>`
     = note: required because it appears within the type `rdkafka::consumer::StreamConsumer`                                                                              
     = note: required because of the requirements on the impl of `std::marker::Send` for `&rdkafka::consumer::StreamConsumer`                                             
     = note: required because it appears within the type `rdkafka::consumer::MessageStream<'_, rdkafka::consumer::DefaultConsumerContext>`                                
     = note: required because it appears within the type `futures::stream::FilterMap<rdkafka::consumer::MessageStream<'_, rdkafka::consumer::DefaultConsumerContext>, [closure@src/main.rs:1755:42: 1763:10]>`
     = note: required because it appears within the type `futures::stream::ForEach<futures::stream::FilterMap<rdkafka::consumer::MessageStream<'_, rdkafka::consumer::DefaultConsumerContext>, [closure@src/main.rs:1755:42: 1763:10]>, [closure@src/main.rs:1763:21: 1829:10 consumer:_, tx:_, input_topic:_], futures::AndThen<futures::MapErr<futures::Receiver<std::string::String>, [closure@src/main.rs:1802:20: 1807:13]>, std::result::Result<(), ()>, [closure@src/main.rs:1813:19: 1826:10 k:_, t1:_]>>`
     = note: required by `tokio::run`

**Comment by surajpayu** [comment]
any plan to do it, it seems that core.run(h) is deprecated and i am not able to run it using tokio::run()

**Comment by toxicafunk** [comment]
@surajpayu you can make it work the way @tyranron suggested or alternatively you can invoke the `asynch_processor`from inside its own thread and then have it exactly as in the repoś example:

```
thread::spawn(move || {
   run_async_processor(brokers, group_id, input_topic);
});
```

this way `current_thread` refers to the spawned thread instead of the `main` thread and you're free to use `tokio::run` or whatever else you use from the main thread.
